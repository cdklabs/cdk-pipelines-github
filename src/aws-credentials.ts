import { awsCredentialStep } from './private/aws-credentials';
import * as github from './workflows-model';

/**
 * AWS credential provider
 */
export abstract class AwsCredentialsProvider {
  public abstract jobPermission(): github.JobPermission;
  public abstract credentialSteps(region: string, assumeRoleArn?: string): github.JobStep[];
}

/**
 * Locations of GitHub Secrets used to authenticate to AWS
 */
export interface GitHubSecretsProviderProps {
  /**
   * @default "AWS_ACCESS_KEY_ID"
   */
  readonly accessKeyId: string;

  /**
   * @default "AWS_SECRET_ACCESS_KEY"
   */
  readonly secretAccessKey: string;

  /**
   * @default - no session token is used
   */
  readonly sessionToken?: string;
}

/**
 * AWS credential provider from GitHub secrets
 */
class GitHubSecretsProvider extends AwsCredentialsProvider {
  private readonly accessKeyId: string;
  private readonly secretAccessKey: string;
  private readonly sessionToken?: string;

  constructor(props?: GitHubSecretsProviderProps) {
    super();
    this.accessKeyId = props?.accessKeyId ?? 'AWS_ACCESS_KEY_ID';
    this.secretAccessKey = props?.secretAccessKey ?? 'AWS_SECRET_ACCESS_KEY';
    this.sessionToken = props?.sessionToken;
  }

  public jobPermission(): github.JobPermission {
    return github.JobPermission.NONE;
  }

  public credentialSteps(region: string, assumeRoleArn?: string): github.JobStep[] {
    return [
      awsCredentialStep('Authenticate Via GitHub Secrets', {
        region,
        accessKeyId: `\${{ secrets.${this.accessKeyId} }}`,
        secretAccessKey: `\${{ secrets.${this.secretAccessKey} }}`,
        ...(this.sessionToken ? {
          sessionToken: `\${{ secrets.${this.sessionToken} }}`,
        } : undefined),
        ...(assumeRoleArn ? {
          roleToAssume: assumeRoleArn,
          roleExternalId: 'Pipeline',
        } : undefined),
      }),
    ];
  }
}

/**
 * Role to assume using OpenId Connect
 */
export interface OpenIdConnectProviderProps {
  /**
   * A role that utilizes the GitHub OIDC Identity Provider in your AWS account.
   *
   * You can create your own role in the console with the necessary trust policy
   * to allow gitHub actions from your gitHub repository to assume the role, or
   * you can utilize the `GitHubActionRole` construct to create a role for you.
   */
  readonly gitHubActionRoleArn: string;
}

/**
 * AWS credential provider from OpenId Connect
 */
class OpenIdConnectProvider extends AwsCredentialsProvider {
  private readonly gitHubActionRoleArn: string;

  constructor(props: OpenIdConnectProviderProps) {
    super();
    this.gitHubActionRoleArn = props.gitHubActionRoleArn;
  }

  public jobPermission(): github.JobPermission {
    return github.JobPermission.WRITE;
  }

  public credentialSteps(region: string, assumeRoleArn?: string): github.JobStep[] {
    function getDeployRole(arn: string) {
      return arn.replace('cfn-exec', 'deploy');
    }

    let steps: github.JobStep[] = [];

    steps.push(
      awsCredentialStep('Authenticate Via OIDC Role', {
        region,
        roleToAssume: this.gitHubActionRoleArn,
      }),
    );

    if (assumeRoleArn) {
      // Result of initial credentials with GitHub Action role are these environment variables
      steps.push(
        awsCredentialStep('Assume CDK Deploy Role', {
          region,
          accessKeyId: '${{ env.AWS_ACCESS_KEY_ID }}',
          secretAccessKey: '${{ env.AWS_SECRET_ACCESS_KEY }}',
          sessionToken: '${{ env.AWS_SESSION_TOKEN }}',
          roleToAssume: getDeployRole(assumeRoleArn),
          roleExternalId: 'Pipeline',
        }),
      );
    }

    return steps;
  }
}

/**
 * Dummy AWS credential provider
 */
class NoCredentialsProvider extends AwsCredentialsProvider {
  public jobPermission(): github.JobPermission {
    return github.JobPermission.NONE;
  }
  public credentialSteps(_region: string, _assumeRoleArn?: string): github.JobStep[] {
    return [];
  }
}

/**
 * Provides AWS credenitals to the pipeline jobs
 */
export class AwsCredentials {
  /**
   * Reference credential secrets to authenticate with AWS. This method assumes
   * that your credentials will be stored as long-lived GitHub Secrets.
   */
  static fromGitHubSecrets(props?: GitHubSecretsProviderProps): AwsCredentialsProvider {
    return new GitHubSecretsProvider(props);
  }

  /**
   * Provide AWS credentials using OpenID Connect.
   */
  static fromOpenIdConnect(props: OpenIdConnectProviderProps): AwsCredentialsProvider {
    return new OpenIdConnectProvider(props);
  }

  /**
   * Don't provide any AWS credentials, use this if runners have preconfigured credentials.
   */
  static runnerHasPreconfiguredCreds(): AwsCredentialsProvider {
    return new NoCredentialsProvider();
  }
}
