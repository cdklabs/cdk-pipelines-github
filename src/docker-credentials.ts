/**
 * Represents a credential used to authenticate to a docker registry.
 * Uses the official Docker Login GitHub Action to authenticate.
 *
 * @see https://github.com/marketplace/actions/docker-login
 */
export class DockerCredential {
  /**
   * Reference credential secrets to authenticate to DockerHub. This method assumes
   * that your credentials will be stored as long-lived GitHub Secrets under the
   * usernameKey and personalAccessTokenKey.
   *
   * The default for usernameKey is `DOCKERHUB_USERNAME`. The default for personalAccessTokenKey
   * is `DOCKERHUB_TOKEN`. If you do not set these values, your credentials should be
   * found in your GitHub Secrets under these default keys.
   */
  public static dockerHub(creds: DockerHubCredentialSecrets = {}): DockerCredential {
    return new DockerCredential(
      'docker',
      undefined,
      creds.usernameKey ?? 'DOCKERHUB_USERNAME',
      creds.personalAccessTokenKey ?? 'DOCKERHUB_TOKEN',
    );
  }

  /**
   * Create a credential for ECR. This method will reuse your AWS credentials to log in to AWS.
   * Your AWS credentials are already used to deploy your CDK stacks. It can be supplied via
   * GitHub Secrets or using an IAM role that trusts the GitHub OIDC identity provider.
   *
   * NOTE - All ECR repositories in the same account and region share a domain name
   * (e.g., 0123456789012.dkr.ecr.eu-west-1.amazonaws.com), and can only have one associated
   * set of credentials (and DockerCredential). Attempting to associate one set of credentials
   * with one ECR repo and another with another ECR repo in the same account and region will
   * result in failures when using these credentials in the pipeline.
   */
  public static ecr(registry: string): DockerCredential {
    return new DockerCredential('ecr', registry);
  }

  /**
   * Create a credential for a custom registry. This method assumes that you will have long-lived
   * GitHub Secrets stored under the usernameKey and passwordKey that will authenticate to the
   * registry you provide.
   *
   * @see https://github.com/marketplace/actions/docker-login
   */
  public static customRegistry(registry: string, creds: ExternalDockerCredentialSecrets): DockerCredential {
    return new DockerCredential('custom', registry, creds.usernameKey, creds.passwordKey);
  }

  private constructor(
    readonly name: string,
    readonly registry?: string,
    readonly usernameKey?: string,
    readonly passwordKey?: string,
  ) {}
}

/**
 * Locations of GitHub Secrets used to authenticate to DockerHub.
 */
export interface DockerHubCredentialSecrets {
  /**
   * The key of the GitHub Secret containing the DockerHub username.
   *
   * @default 'DOCKERHUB_USERNAME'
   */
  readonly usernameKey?: string;

  /**
   * The key of the GitHub Secret containing the DockerHub personal access token.
   *
   * @default 'DOCKERHUB_TOKEN'
   */
  readonly personalAccessTokenKey?: string;
}

/**
 * Generic structure to supply the locations of GitHub Secrets used to authenticate
 * to a docker registry.
 */
export interface ExternalDockerCredentialSecrets {
  /**
   * The key of the GitHub Secret containing your registry username.
   */
  readonly usernameKey: string;

  /**
   * The key of the GitHub Secret containing your registry password.
   */
  readonly passwordKey: string;
}
