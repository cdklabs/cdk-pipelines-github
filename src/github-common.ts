import { AddStageOpts } from 'aws-cdk-lib/pipelines';
import { JobSettings } from './pipeline';
import { StackCapabilities } from './stage-options';

/**
 * Common properties to extend both StageProps and AddStageOpts
 */

export interface GitHubCommonProps {
  /**
   * Run the stage in a specific GitHub Environment. If specified,
   * any protection rules configured for the environment must pass
   * before the job is set to a runner. For example, if the environment
   * has a manual approval rule configured, then the workflow will
   * wait for the approval before sending the job to the runner.
   *
   * Running a workflow that references an environment that does not
   * exist will create an environment with the referenced name.
   * @see https://docs.github.com/en/actions/deployment/targeting-different-environments/using-environments-for-deployment
   *
   * @default - no GitHub environment
   */
  readonly gitHubEnvironment?: string;

  /**
   * In some cases, you must explicitly acknowledge that your CloudFormation
   * stack template contains certain capabilities in order for CloudFormation
   * to create the stack.
   *
   * If insufficiently specified, CloudFormation returns an `InsufficientCapabilities`
   * error.
   *
   * @default ['CAPABILITY_IAM']
   */
  readonly stackCapabilities?: StackCapabilities[];

  /**
   * Job level settings that will be applied to all jobs in the stage.
   * Currently the only valid setting is 'if'.
   */
  readonly jobSettings?: JobSettings;
}

/**
 * Options to pass to `addStageWithGitHubOpts`.
 */
export interface AddGitHubStageOptions extends AddStageOpts, GitHubCommonProps {}
