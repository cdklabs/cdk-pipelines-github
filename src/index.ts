export * from './pipeline';
export * from './workflows-model';
export * from './oidc-provider';
export * from './docker-credentials';
export * from './stage-options';
