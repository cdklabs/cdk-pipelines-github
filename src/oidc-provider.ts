import { Aws, CfnOutput } from 'aws-cdk-lib';
import * as iam from 'aws-cdk-lib/aws-iam';
import { Construct } from 'constructs';

/**
 * GitHub OIDC thumbprints updated 2023-07-27
 *
 * https://github.blog/changelog/2023-06-27-github-actions-update-on-oidc-integration-with-aws/
 */
const GITHUB_OIDC_THUMBPRINTS = [
  '6938fd4d98bab03faadb97b34396831e3780aea1',
  '1c58a3a8518e8759bf075b76b750d4f2df264fcd',
];

/**
 * Properties for the GitHubActionRole construct.
 */
export interface GitHubActionRoleProps {
  /**
   * A list of GitHub repositories you want to be able to access the IAM role.
   * Each entry should be your GitHub username and repository passed in as a
   * single string.
   * An entry `owner/repo` is equivalent to the subjectClaim `repo:owner/repo:*`.
   *
   * For example, `['owner/repo1', 'owner/repo2'].
   */
  readonly repos?: string[];

  /**
   * A list of subject claims allowed to access the IAM role.
   * See https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/about-security-hardening-with-openid-connect
   * A subject claim can include `*` and `?` wildcards according to the `StringLike`
   * condition operator.
   *
   * For example, `['repo:owner/repo1:ref:refs/heads/branch1', 'repo:owner/repo1:environment:prod']`
   */
  readonly subjectClaims?: string[];

  /**
   * The name of the Oidc role.
   *
   * @default 'GitHubActionRole'
   */
  readonly roleName?: string;

  /**
   * The GitHub OpenId Connect Provider. Must have provider url
   * `https://token.actions.githubusercontent.com`. The audience must be
   * `sts:amazonaws.com`.
   *
   * Only one such provider can be defined per account, so if you already
   * have a provider with the same url, a new provider cannot be created for you.
   *
   * @default - a provider is created for you.
   */
  readonly provider?: iam.IOpenIdConnectProvider;

  /**
   * Thumbprints of GitHub's certificates
   *
   * Every time GitHub rotates their certificates, this value will need to be updated.
   *
   * Default value is up-to-date to June 27, 2023 as per
   * https://github.blog/changelog/2023-06-27-github-actions-update-on-oidc-integration-with-aws/
   *
   * @default - Use built-in keys
   */
  readonly thumbprints?: string[];
}

/**
 * Creates or references a GitHub OIDC provider and accompanying role that trusts the provider.
 * This role can be used to authenticate against AWS instead of using long-lived AWS user credentials
 * stored in GitHub secrets.
 *
 * You can do this manually in the console, or create a separate stack that uses this construct.
 * You must `cdk deploy` once (with your normal AWS credentials) to have this role created for you.
 *
 * You can then make note of the role arn in the stack output and send it into the Github Workflow app via
 * the `gitHubActionRoleArn` property. The role arn will be `arn:<partition>:iam::<accountId>:role/GithubActionRole`.
 *
 * @see https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-amazon-web-services
 */
export class GitHubActionRole extends Construct {
  /**
   * Reference an existing GitHub Actions provider.
   * You do not need to pass in an arn because the arn for such
   * a provider is always the same.
   */
  public static existingGitHubActionsProvider(scope: Construct): iam.IOpenIdConnectProvider {
    return iam.OpenIdConnectProvider.fromOpenIdConnectProviderArn(
      scope,
      'GitHubActionProvider',
      `arn:${Aws.PARTITION}:iam::${Aws.ACCOUNT_ID}:oidc-provider/token.actions.githubusercontent.com`,
    );
  }

  /**
   * The role that gets created.
   *
   * You should use the arn of this role as input to the `gitHubActionRoleArn`
   * property in your GitHub Workflow app.
   */
  public readonly role: iam.IRole;

  constructor(scope: Construct, id: string, props: GitHubActionRoleProps) {
    super(scope, id);

    const rawEndpoint = 'token.actions.githubusercontent.com';
    const providerUrl = `https://${rawEndpoint}`;

    // uses the given provider or creates a new one.
    const provider = props.provider ?? new iam.OpenIdConnectProvider(this, 'github-provider', {
      url: providerUrl,
      clientIds: ['sts.amazonaws.com'],
      thumbprints: props.thumbprints ?? GITHUB_OIDC_THUMBPRINTS,
    });

    // create a role that references the provider as a trusted entity
    const principal = new iam.FederatedPrincipal(
      provider.openIdConnectProviderArn,
      {
        StringLike: {
          [`${rawEndpoint}:sub`]: formatRepos(props.repos ?? []).concat(props.subjectClaims ?? []),
        },
      },
      'sts:AssumeRoleWithWebIdentity',
    );

    // permit this role from assuming all of the CDK bootstrap roles
    const oidcPolicyStatement = new iam.PolicyStatement({
      effect: iam.Effect.ALLOW,
      actions: ['sts:AssumeRole'],
      resources: ['*'],
      conditions: {
        'ForAnyValue:StringEquals': {
          'iam:ResourceTag/aws-cdk:bootstrap-role': [
            'deploy',
            'lookup',
            'file-publishing',
            'image-publishing',
          ],
        },
      },
    });

    // permit this role from accessing ecr repositories for docker assets
    const ecrPolicyStatement = new iam.PolicyStatement({
      effect: iam.Effect.ALLOW,
      actions: ['ecr:GetAuthorizationToken'],
      resources: ['*'],
    });

    this.role = new iam.Role(this, 'github-action-role', {
      roleName: props.roleName ?? 'GitHubActionRole',
      assumedBy: principal,
      inlinePolicies: {
        AssumeBootstrapRoles: new iam.PolicyDocument({
          statements: [oidcPolicyStatement, ecrPolicyStatement],
        }),
      },
    });

    // show the role arn in the stack output
    new CfnOutput(this, 'roleArn', {
      value: this.role.roleArn,
    });
  }
}

function formatRepos(repos: string[]) {
  const formattedRepos = [];
  for (const repo of repos) {
    formattedRepos.push(`repo:${repo}:*`);
  }
  return formattedRepos;
}
