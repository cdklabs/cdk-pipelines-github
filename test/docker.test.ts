import * as fs from 'fs';
import * as os from 'os';
import { join } from 'path';
import { App, Stack, Stage, StageProps } from 'aws-cdk-lib';
import * as codebuild from 'aws-cdk-lib/aws-codebuild';
import { ShellStep } from 'aws-cdk-lib/pipelines';
import { DockerCredential, GitHubWorkflow } from '../src';
import { TestApp } from './testutil';

const dockers = join(__dirname, 'demo-image');

describe('correct format for docker credentials:', () => {
  test('ECR', () => {
    const github = createDockerGithubWorkflow([DockerCredential.ecr('000000000000.dkr.ecr.us-east-1.amazonaws.com')]);
    expect(fs.readFileSync(github.workflowPath, 'utf-8')).toMatchSnapshot();
  });

  test('DockerHub', () => {
    const github = createDockerGithubWorkflow([DockerCredential.dockerHub()]);
    expect(fs.readFileSync(github.workflowPath, 'utf-8')).toMatchSnapshot();
  });

  test('custom registry', () => {
    const github = createDockerGithubWorkflow([
      DockerCredential.customRegistry('registry', {
        usernameKey: 'CUSTOM_USERNAME',
        passwordKey: 'CUSTOM_PASSWORD',
      }),
    ]);
    expect(fs.readFileSync(github.workflowPath, 'utf-8')).toMatchSnapshot();
  });

  test('multiple credentials', () => {
    const github = createDockerGithubWorkflow([
      DockerCredential.dockerHub(),
      DockerCredential.ecr('000000000000.dkr.ecr.us-east-1.amazonaws.com'),
    ]);

    expect(fs.readFileSync(github.workflowPath, 'utf-8')).toMatchSnapshot();
  });
});

function createDockerGithubWorkflow(dockerCredentials: DockerCredential[]) {
  const tempOutDir = 'temp.out';
  const app = new TestApp({
    outdir: tempOutDir,
  });

  const github = new GitHubWorkflow(app, 'Pipeline', {
    workflowPath: `${mkoutdir()}/.github/workflows/deploy.yml`,
    synth: new ShellStep('Build', {
      installCommands: ['yarn'],
      commands: ['yarn build'],
    }),
    dockerCredentials,
  });

  github.addStage(new MyDockerStage(app, 'MyStage', {
    env: {
      account: '000000000000',
      region: 'us-east-1',
    },
  }));

  app.synth();

  return github;
}

class MyDockerStage extends Stage {
  constructor(scope: App, id: string, props: StageProps) {
    super(scope, id, props);

    const buildStack = new Stack(this, 'BuildStack');

    new codebuild.Project(buildStack, 'MyProject', {
      buildSpec: codebuild.BuildSpec.fromObject({
        version: '0.2',
        phases: {
          build: {
            commands: ['ls'],
          },
        },
      }),
      grantReportGroupPermissions: false,
      environment: {
        buildImage: codebuild.LinuxBuildImage.fromAsset(buildStack, 'MyImage', {
          directory: dockers,
        }),
      },
    });
  }
}

function mkoutdir() {
  return fs.mkdtempSync(join(os.tmpdir(), 'cdk-pipelines-github-'));
}