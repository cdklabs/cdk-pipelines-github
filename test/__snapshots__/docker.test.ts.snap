// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`correct format for docker credentials: DockerHub 1`] = `
"name: deploy
on:
  push:
    branches:
      - main
  workflow_dispatch: {}
jobs:
  Build-Build:
    name: Synthesize
    permissions:
      contents: read
    runs-on: ubuntu-latest
    needs: []
    env: {}
    container: null
    steps:
      - name: Checkout
        uses: actions/checkout@v2
      - name: Install
        run: yarn
      - name: Build
        run: yarn build
      - name: Upload cdk.out
        uses: actions/upload-artifact@v2.1.1
        with:
          name: cdk.out
          path: cdk.out
  Assets-DockerAsset1:
    name: Publish Assets Assets-DockerAsset1
    needs:
      - Build-Build
    permissions:
      contents: read
    runs-on: ubuntu-latest
    steps:
      - name: Download cdk.out
        uses: actions/download-artifact@v2
        with:
          name: cdk.out
          path: temp.out
      - name: Install
        run: npm install --no-save cdk-assets
      - uses: aws-actions/configure-aws-credentials@v1
        with:
          aws-access-key-id: \${{ secrets.AWS_ACCESS_KEY_ID }}
          aws-secret-access-key: \${{ secrets.AWS_SECRET_ACCESS_KEY }}
          aws-region: us-west-2
          role-skip-session-tagging: true
          role-duration-seconds: 1800
      - uses: docker/login-action@v1
        with:
          username: \${{ secrets.DOCKERHUB_USERNAME }}
          password: \${{ secrets.DOCKERHUB_TOKEN }}
      - name: Publish Assets-DockerAsset1
        run: /bin/bash ./cdk.out/publish-Assets-DockerAsset1-step.sh
  Assets-FileAsset1:
    name: Publish Assets Assets-FileAsset1
    needs:
      - Build-Build
    permissions:
      contents: read
    runs-on: ubuntu-latest
    steps:
      - name: Download cdk.out
        uses: actions/download-artifact@v2
        with:
          name: cdk.out
          path: temp.out
      - name: Install
        run: npm install --no-save cdk-assets
      - uses: aws-actions/configure-aws-credentials@v1
        with:
          aws-access-key-id: \${{ secrets.AWS_ACCESS_KEY_ID }}
          aws-secret-access-key: \${{ secrets.AWS_SECRET_ACCESS_KEY }}
          aws-region: us-west-2
          role-skip-session-tagging: true
          role-duration-seconds: 1800
      - name: Publish Assets-FileAsset1
        run: /bin/bash ./cdk.out/publish-Assets-FileAsset1-step.sh
  MyStage-BuildStack-Deploy:
    name: Deploy MyStageBuildStackEE7613FD
    permissions:
      contents: none
    needs:
      - Build-Build
      - Assets-FileAsset1
      - Assets-DockerAsset1
      - Build-Build
    runs-on: ubuntu-latest
    steps:
      - uses: aws-actions/configure-aws-credentials@v1
        with:
          aws-access-key-id: \${{ secrets.AWS_ACCESS_KEY_ID }}
          aws-secret-access-key: \${{ secrets.AWS_SECRET_ACCESS_KEY }}
          aws-region: us-east-1
          role-skip-session-tagging: true
          role-duration-seconds: 1800
          role-to-assume: arn:aws:iam::000000000000:role/cdk-hnb659fds-deploy-role-000000000000-us-east-1
          role-external-id: Pipeline
      - id: Deploy
        uses: aws-actions/aws-cloudformation-github-deploy@v1
        with:
          name: MyStage-BuildStack
          template: https://cdk-hnb659fds-assets-000000000000-us-east-1.s3.us-east-1.amazonaws.com/bf3395b0ca977c5e3fcf8e3905a8a6981b81d7356e390da732ec29f385bb7ec0.json
          no-fail-on-empty-changeset: \\"1\\"
          role-arn: arn:aws:iam::000000000000:role/cdk-hnb659fds-cfn-exec-role-000000000000-us-east-1
"
`;

exports[`correct format for docker credentials: ECR 1`] = `
"name: deploy
on:
  push:
    branches:
      - main
  workflow_dispatch: {}
jobs:
  Build-Build:
    name: Synthesize
    permissions:
      contents: read
    runs-on: ubuntu-latest
    needs: []
    env: {}
    container: null
    steps:
      - name: Checkout
        uses: actions/checkout@v2
      - name: Install
        run: yarn
      - name: Build
        run: yarn build
      - name: Upload cdk.out
        uses: actions/upload-artifact@v2.1.1
        with:
          name: cdk.out
          path: cdk.out
  Assets-DockerAsset1:
    name: Publish Assets Assets-DockerAsset1
    needs:
      - Build-Build
    permissions:
      contents: read
    runs-on: ubuntu-latest
    steps:
      - name: Download cdk.out
        uses: actions/download-artifact@v2
        with:
          name: cdk.out
          path: temp.out
      - name: Install
        run: npm install --no-save cdk-assets
      - uses: aws-actions/configure-aws-credentials@v1
        with:
          aws-access-key-id: \${{ secrets.AWS_ACCESS_KEY_ID }}
          aws-secret-access-key: \${{ secrets.AWS_SECRET_ACCESS_KEY }}
          aws-region: us-west-2
          role-skip-session-tagging: true
          role-duration-seconds: 1800
      - uses: docker/login-action@v1
        with:
          registry: 000000000000.dkr.ecr.us-east-1.amazonaws.com
      - name: Publish Assets-DockerAsset1
        run: /bin/bash ./cdk.out/publish-Assets-DockerAsset1-step.sh
  Assets-FileAsset1:
    name: Publish Assets Assets-FileAsset1
    needs:
      - Build-Build
    permissions:
      contents: read
    runs-on: ubuntu-latest
    steps:
      - name: Download cdk.out
        uses: actions/download-artifact@v2
        with:
          name: cdk.out
          path: temp.out
      - name: Install
        run: npm install --no-save cdk-assets
      - uses: aws-actions/configure-aws-credentials@v1
        with:
          aws-access-key-id: \${{ secrets.AWS_ACCESS_KEY_ID }}
          aws-secret-access-key: \${{ secrets.AWS_SECRET_ACCESS_KEY }}
          aws-region: us-west-2
          role-skip-session-tagging: true
          role-duration-seconds: 1800
      - name: Publish Assets-FileAsset1
        run: /bin/bash ./cdk.out/publish-Assets-FileAsset1-step.sh
  MyStage-BuildStack-Deploy:
    name: Deploy MyStageBuildStackEE7613FD
    permissions:
      contents: none
    needs:
      - Build-Build
      - Assets-FileAsset1
      - Assets-DockerAsset1
      - Build-Build
    runs-on: ubuntu-latest
    steps:
      - uses: aws-actions/configure-aws-credentials@v1
        with:
          aws-access-key-id: \${{ secrets.AWS_ACCESS_KEY_ID }}
          aws-secret-access-key: \${{ secrets.AWS_SECRET_ACCESS_KEY }}
          aws-region: us-east-1
          role-skip-session-tagging: true
          role-duration-seconds: 1800
          role-to-assume: arn:aws:iam::000000000000:role/cdk-hnb659fds-deploy-role-000000000000-us-east-1
          role-external-id: Pipeline
      - id: Deploy
        uses: aws-actions/aws-cloudformation-github-deploy@v1
        with:
          name: MyStage-BuildStack
          template: https://cdk-hnb659fds-assets-000000000000-us-east-1.s3.us-east-1.amazonaws.com/bf3395b0ca977c5e3fcf8e3905a8a6981b81d7356e390da732ec29f385bb7ec0.json
          no-fail-on-empty-changeset: \\"1\\"
          role-arn: arn:aws:iam::000000000000:role/cdk-hnb659fds-cfn-exec-role-000000000000-us-east-1
"
`;

exports[`correct format for docker credentials: custom registry 1`] = `
"name: deploy
on:
  push:
    branches:
      - main
  workflow_dispatch: {}
jobs:
  Build-Build:
    name: Synthesize
    permissions:
      contents: read
    runs-on: ubuntu-latest
    needs: []
    env: {}
    container: null
    steps:
      - name: Checkout
        uses: actions/checkout@v2
      - name: Install
        run: yarn
      - name: Build
        run: yarn build
      - name: Upload cdk.out
        uses: actions/upload-artifact@v2.1.1
        with:
          name: cdk.out
          path: cdk.out
  Assets-DockerAsset1:
    name: Publish Assets Assets-DockerAsset1
    needs:
      - Build-Build
    permissions:
      contents: read
    runs-on: ubuntu-latest
    steps:
      - name: Download cdk.out
        uses: actions/download-artifact@v2
        with:
          name: cdk.out
          path: temp.out
      - name: Install
        run: npm install --no-save cdk-assets
      - uses: aws-actions/configure-aws-credentials@v1
        with:
          aws-access-key-id: \${{ secrets.AWS_ACCESS_KEY_ID }}
          aws-secret-access-key: \${{ secrets.AWS_SECRET_ACCESS_KEY }}
          aws-region: us-west-2
          role-skip-session-tagging: true
          role-duration-seconds: 1800
      - uses: docker/login-action@v1
        with:
          registry: registry
          username: \${{ secrets.CUSTOM_USERNAME }}
          password: \${{ secrets.CUSTOM_PASSWORD }}
      - name: Publish Assets-DockerAsset1
        run: /bin/bash ./cdk.out/publish-Assets-DockerAsset1-step.sh
  Assets-FileAsset1:
    name: Publish Assets Assets-FileAsset1
    needs:
      - Build-Build
    permissions:
      contents: read
    runs-on: ubuntu-latest
    steps:
      - name: Download cdk.out
        uses: actions/download-artifact@v2
        with:
          name: cdk.out
          path: temp.out
      - name: Install
        run: npm install --no-save cdk-assets
      - uses: aws-actions/configure-aws-credentials@v1
        with:
          aws-access-key-id: \${{ secrets.AWS_ACCESS_KEY_ID }}
          aws-secret-access-key: \${{ secrets.AWS_SECRET_ACCESS_KEY }}
          aws-region: us-west-2
          role-skip-session-tagging: true
          role-duration-seconds: 1800
      - name: Publish Assets-FileAsset1
        run: /bin/bash ./cdk.out/publish-Assets-FileAsset1-step.sh
  MyStage-BuildStack-Deploy:
    name: Deploy MyStageBuildStackEE7613FD
    permissions:
      contents: none
    needs:
      - Build-Build
      - Assets-FileAsset1
      - Assets-DockerAsset1
      - Build-Build
    runs-on: ubuntu-latest
    steps:
      - uses: aws-actions/configure-aws-credentials@v1
        with:
          aws-access-key-id: \${{ secrets.AWS_ACCESS_KEY_ID }}
          aws-secret-access-key: \${{ secrets.AWS_SECRET_ACCESS_KEY }}
          aws-region: us-east-1
          role-skip-session-tagging: true
          role-duration-seconds: 1800
          role-to-assume: arn:aws:iam::000000000000:role/cdk-hnb659fds-deploy-role-000000000000-us-east-1
          role-external-id: Pipeline
      - id: Deploy
        uses: aws-actions/aws-cloudformation-github-deploy@v1
        with:
          name: MyStage-BuildStack
          template: https://cdk-hnb659fds-assets-000000000000-us-east-1.s3.us-east-1.amazonaws.com/bf3395b0ca977c5e3fcf8e3905a8a6981b81d7356e390da732ec29f385bb7ec0.json
          no-fail-on-empty-changeset: \\"1\\"
          role-arn: arn:aws:iam::000000000000:role/cdk-hnb659fds-cfn-exec-role-000000000000-us-east-1
"
`;

exports[`correct format for docker credentials: multiple credentials 1`] = `
"name: deploy
on:
  push:
    branches:
      - main
  workflow_dispatch: {}
jobs:
  Build-Build:
    name: Synthesize
    permissions:
      contents: read
    runs-on: ubuntu-latest
    needs: []
    env: {}
    container: null
    steps:
      - name: Checkout
        uses: actions/checkout@v2
      - name: Install
        run: yarn
      - name: Build
        run: yarn build
      - name: Upload cdk.out
        uses: actions/upload-artifact@v2.1.1
        with:
          name: cdk.out
          path: cdk.out
  Assets-DockerAsset1:
    name: Publish Assets Assets-DockerAsset1
    needs:
      - Build-Build
    permissions:
      contents: read
    runs-on: ubuntu-latest
    steps:
      - name: Download cdk.out
        uses: actions/download-artifact@v2
        with:
          name: cdk.out
          path: temp.out
      - name: Install
        run: npm install --no-save cdk-assets
      - uses: aws-actions/configure-aws-credentials@v1
        with:
          aws-access-key-id: \${{ secrets.AWS_ACCESS_KEY_ID }}
          aws-secret-access-key: \${{ secrets.AWS_SECRET_ACCESS_KEY }}
          aws-region: us-west-2
          role-skip-session-tagging: true
          role-duration-seconds: 1800
      - uses: docker/login-action@v1
        with:
          username: \${{ secrets.DOCKERHUB_USERNAME }}
          password: \${{ secrets.DOCKERHUB_TOKEN }}
      - uses: docker/login-action@v1
        with:
          registry: 000000000000.dkr.ecr.us-east-1.amazonaws.com
      - name: Publish Assets-DockerAsset1
        run: /bin/bash ./cdk.out/publish-Assets-DockerAsset1-step.sh
  Assets-FileAsset1:
    name: Publish Assets Assets-FileAsset1
    needs:
      - Build-Build
    permissions:
      contents: read
    runs-on: ubuntu-latest
    steps:
      - name: Download cdk.out
        uses: actions/download-artifact@v2
        with:
          name: cdk.out
          path: temp.out
      - name: Install
        run: npm install --no-save cdk-assets
      - uses: aws-actions/configure-aws-credentials@v1
        with:
          aws-access-key-id: \${{ secrets.AWS_ACCESS_KEY_ID }}
          aws-secret-access-key: \${{ secrets.AWS_SECRET_ACCESS_KEY }}
          aws-region: us-west-2
          role-skip-session-tagging: true
          role-duration-seconds: 1800
      - name: Publish Assets-FileAsset1
        run: /bin/bash ./cdk.out/publish-Assets-FileAsset1-step.sh
  MyStage-BuildStack-Deploy:
    name: Deploy MyStageBuildStackEE7613FD
    permissions:
      contents: none
    needs:
      - Build-Build
      - Assets-FileAsset1
      - Assets-DockerAsset1
      - Build-Build
    runs-on: ubuntu-latest
    steps:
      - uses: aws-actions/configure-aws-credentials@v1
        with:
          aws-access-key-id: \${{ secrets.AWS_ACCESS_KEY_ID }}
          aws-secret-access-key: \${{ secrets.AWS_SECRET_ACCESS_KEY }}
          aws-region: us-east-1
          role-skip-session-tagging: true
          role-duration-seconds: 1800
          role-to-assume: arn:aws:iam::000000000000:role/cdk-hnb659fds-deploy-role-000000000000-us-east-1
          role-external-id: Pipeline
      - id: Deploy
        uses: aws-actions/aws-cloudformation-github-deploy@v1
        with:
          name: MyStage-BuildStack
          template: https://cdk-hnb659fds-assets-000000000000-us-east-1.s3.us-east-1.amazonaws.com/bf3395b0ca977c5e3fcf8e3905a8a6981b81d7356e390da732ec29f385bb7ec0.json
          no-fail-on-empty-changeset: \\"1\\"
          role-arn: arn:aws:iam::000000000000:role/cdk-hnb659fds-cfn-exec-role-000000000000-us-east-1
"
`;
